src/persist-types/src/stats.proto

```diff
@@ -0,0 +1,67 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+syntax = "proto3";
+
+package mz_persist_types.stats;
+
+message ProtoStructStats {
+    uint64 len = 1;
+    map<string, ProtoDynStats> cols = 2;
+}
+
+message ProtoDynStats {
+    // ProtoOptionStats could instead be have a `ProtoDynStats some` field and
+    // be in the oneof, but that's unnecesary (we don't need
+    // `OptionStats<OptionStats<T>>`) and causes some problems (OptionStats
+    // would need to represent `some` as a `Box<dyn DynStats>` instead of a `T`,
+    // which is far less ergonomic).
```

Comment:
> is it to replace ProtoOptionStats option with uint64 null_count?

Yes!

> how does that reduce invalid states?

Currently three interesting states for this field: `ProtoOptionStats` not present, `ProtoOptionStats` present but is  zero / default / unset value, `ProtoOptionStats` present with a nondefault value. The first is invalid for optional stats, and the latter two are invalid for non-optional stats.

The `null_count` idea has two states: field zero / unset or not. Neither is invalid for optional stats, only the latter is invalid for non-optional stats. So 1/4 invalid instead of 3/6!

> also, if that is the suggestion, I don't think it works [...]

I was having trouble thinking why we'd care about the distinction, but I think maybe the idea is that we're expecting to be able to tell whether this is an optional or non-optional stats purely by the proto representation? (vs. needing the schema information in hand to make sense of it.) Which seems reasonable if so.

Reply:
ah, yeah, at the moment, `<Box<dyn DynStats>>::from_proto` is how we deserialize this and it has to figure out the structure purely from the proto representation. now that you mention it, maybe it is better to be given the desired rust type and then figure out how to parse it from the proto, I'll play with that tomorrow

Reply:
(and the invalid states thing makes sense now, thanks!)

Reply:
I played with this for a bit locally and couldn't manage anything better than what's here now

the `null_count` idea doesn't work because this needs to be self-describing. my attempts to remove that requirement kept getting tripped up by OptionStats having `some: T` instead of `some: Box<dyn DynStats>`, which I think is worth keeping